llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -181,8 +181,12 @@ Function *ESIMDLowerVecArgPass::rewriteFunc(Function &F) {
 
   llvm::CloneFunctionInto(NF, &F, VMap, F.getSubprogram() != nullptr, Returns);
 
+  // insert bitcasts in new function only if its a definition
   for (auto &B : BitCasts) {
-    NF->begin()->getInstList().push_front(B);
+    if (!F.isDeclaration())
+      NF->begin()->getInstList().push_front(B);
+    else
+      delete B;
```

Comment:
I think you can put the `if (!F.isDeclaration())` check in the previous loop where you populate the `BitCasts` vector. This way you don't need to `delete B`.

Reply:
@dendibakh - please refer to discussion above discussing exactly this.

Reply:
@pratikashar - ah, OK.